osu.Game/Screens/Edit/EditorCommandHandler.cs

```diff
@@ -0,0 +1,312 @@
+            }
+
+            undoStack.Push(undoTransaction);
+
+            historyChanged();
+
+            return true;
+        }
+
+        /// <summary>
+        /// Reverts any changes that have been made since the last commit.
+        /// Returns false if there are no uncommitted changes.
+        /// </summary>
+        public bool RevertUncommitedChanges()
+        {
+            if (!HasUncommittedChanges)
+                return false;
+
+            revertTransaction(currentTransaction);
+
+            currentTransaction = new Transaction();
+
+            return true;
+        }
+
+        private void revertTransaction(Transaction transaction)
+        {
+            // We are navigating history so we don't want to write a new state.
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = true;
+
+            editorBeatmap.BeginChange();
+
+            foreach (var command in transaction.UndoCommands)
+                apply(command);
+
+            editorBeatmap.EndChange();
+
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = false;
+        }
+
+        private void historyChanged()
+        {
+            CanUndo.Value = undoStack.Count > 0;
+            CanRedo.Value = redoStack.Count > 0;
+        }
+
+        private void apply(IEditorCommand command)
+        {
+            command.Apply();
+            CommandApplied?.Invoke(command);
+        }
+
+        private void record(IEditorCommand command)
+        {
+            var reverse = command.CreateUndo();
+
+            currentTransaction.Add(reverse);
+        }
+
+        private readonly struct Transaction
+        {
+            public readonly bool IsChangeHandlerTransaction;
+
+            public Transaction()
+            {
+                undoCommands = new List<IEditorCommand>();
+            }
+
+            public Transaction(bool isChangeHandlerTransaction = false)
+            {
+                IsChangeHandlerTransaction = isChangeHandlerTransaction;
+                undoCommands = new List<IEditorCommand>();
+            }
+
+            private Transaction(List<IEditorCommand> undoCommands, bool isChangeHandlerTransaction = false)
+            {
+                IsChangeHandlerTransaction = isChangeHandlerTransaction;
+                this.undoCommands = undoCommands;
+            }
+
+            private readonly List<IEditorCommand> undoCommands;
+
+            /// <summary>
+            /// The commands to undo the given transaction.
+            /// Stored in reverse order of original commands to match execution order when undoing.
+            /// </summary>
+            public IReadOnlyList<IEditorCommand> UndoCommands => undoCommands;
+
+            public void Add(IEditorCommand command)
+            {
+                if (command is IMergeableCommand mergeable)
+                {
+                    for (int i = 0; i < undoCommands.Count; i++)
```

Comment:
Mergable commands work under the assumption that they represent atomic changes to the "state" (the parts of the beatmap data structure that act as a source of truth for anything else) without side effects.
Under that assumption the order in this case should have no effect on the final outcome.

Reply:
Here's an even simpler example that breaks the merging logic.

```cs
Proxy.SetPosition(v1);
Proxy.SetX(x);
Proxy.SetPosition(v2);
```

If you merge the two SetPositions over the SetX, you'll lose correctness. Since X and Position are not independent.

Reply:
You're right that would break things. This is one of the primary reasons why `IHasMutablePosition` does not extend `IHasMutableXPosition` or `IHasMutableYPosition` to reduce the risk of having these kinds of situations.

Reply:
Undo and redo still work correctly in this last example. We'd end up with a redo stack that looks like this
```cs
Proxy.SetPosition((v1.X, v2.Y));
Proxy.SetX(v2.X);
```